### PR TITLE
[Merged by Bors] - feat: port Logic.Encodable.Basic

### PR DESCRIPTION
**Pull request by arienmalec**
Port of logic.encodable.basic

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Logic/Encodable/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,699 @@
...
+#align encodable.decode_sum_val Encodable.decode_sum_val
+
+end Sum
+
+instance Bool.encodable : Encodable Bool :=
+  ofEquiv (Sum Unit Unit) Equiv.boolEquivPUnitSumPUnit
+#align bool.encodable Encodable.Bool.encodable
+
+@[simp]
+theorem encode_tt : encode true = 1 :=
```
Comment (jcommelin) [inline Mathlib/Logic/Encodable/Basic.lean:311]:
```suggestion
theorem encode_true : encode true = 1 :=
```

File: Mathlib/Logic/Encodable/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,699 @@
...
+  rfl
+#align encodable.encode_tt Encodable.encode_tt
+
+@[simp]
+theorem encode_ff : encode false = 0 :=
```
Comment (jcommelin) [inline Mathlib/Logic/Encodable/Basic.lean:316]:
```suggestion
theorem encode_false : encode false = 0 :=
```

**Comment by jcommelin** [comment]
Apart from the naming issue flagged above, LGTM.

File: Mathlib/Logic/Encodable/Basic.lean [hunk 3]
```diff
@@ -0,0 +1,699 @@
...
+    Encodable β :=
+  ofLeftInjection f (some ∘ finv) fun b => congr_arg some (linv b)
+#align encodable.of_left_inverse Encodable.ofLeftInverse
+
+/-- Encodability is preserved by equivalence. -/
+def ofEquiv (α) [Encodable α] (e : β ≃ α) : Encodable β :=
+  ofLeftInverse e e.symm e.left_inv
+#align encodable.of_equiv Encodable.ofEquiv
+
+-- Porting note: removing @[simp], too powerful
```
Comment (jcommelin) [inline Mathlib/Logic/Encodable/Basic.lean:110]: This is a `rfl`-lemma. Can `dsimp` prove this? If so, mention it in the porting note.
```suggestion
-- @[simp] -- Porting note: dsimp can prove this
```
If not please use
```suggestion
@[simp, nolint simpNF] -- Porting note: dsimp can not prove this
```

Comment (arienmalec) [reply Mathlib/Logic/Encodable/Basic.lean:110]: @ChrisHughes24  -- deferring to you on this

Comment (arienmalec) [reply Mathlib/Logic/Encodable/Basic.lean:110]: Or @jcommelin  -- if I replace `rfl` by `by dsimp` there's no close -- is that the test you were looking for?

**Comment by jcommelin** [comment]
lint is failing

**Comment by arienmalec** [comment]
> lint is failing

Posted to Zulip.

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**